code/game/objects/items/devices/gps.dm

```diff
@@ -1,28 +1,423 @@
+		moved_event.register(holder, src, .proc/update_compass)
+		dir_set_event.register(holder, src, .proc/update_compass)
+
+	if(holder && tracking)
+		if(!is_in_processing_list)
+			START_PROCESSING(SSobj, src)
+			is_in_processing_list = TRUE
+		if(holder.client)
+			if(check_visible_to_holder())
+				holder.client.screen |= compass
+			else
+				holder.client.screen -= compass
+	else
+		STOP_PROCESSING(SSobj, src)
+		is_in_processing_list = FALSE
+		if(holder?.client)
+			holder.client.screen -= compass
+
+/obj/item/gps/pickup()
+	. = ..()
+	update_holder()
+
+/obj/item/gps/equipped_robot()
+	. = ..()
+	update_holder()
+
+/obj/item/gps/equipped()
+	. = ..()
+	update_holder()
+
+/obj/item/gps/dropped()
+	. = ..()
+	update_holder()
+
+/obj/item/gps/Process()
+	if(!tracking)
+		is_in_processing_list = FALSE
+		return PROCESS_KILL
+	update_holder()
+	if(holder)
+		update_compass(TRUE)
+
+/obj/item/gps/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	is_in_processing_list = FALSE
+	global.all_gps_units -= src
+	. = ..()
+	update_holder()
+	QDEL_NULL(compass)
+
+/obj/item/gps/proc/can_track(var/obj/item/gps/other, var/reachable_z_levels)
+	if(!other.tracking || other.emped || other.hide_signal)
+		return FALSE
+	reachable_z_levels = reachable_z_levels || list()
+	var/turf/origin = get_turf(src)
+	var/turf/target = get_turf(other)
+	if(!istype(origin) || !istype(target))
+		return FALSE
+	if(origin.z == target.z)
+		return TRUE
+	if(local_mode)
+		return FALSE
+
+	if(long_range)
+		reachable_z_levels |= (global.using_map.station_levels|global.using_map.contact_levels|global.using_map.player_levels)
+	else
+		reachable_z_levels |= GetConnectedZlevels(origin.z)
+	return (target.z in reachable_z_levels)
+
+/obj/item/gps/proc/update_compass(var/update_compass_icon)
+
+	compass.hide_waypoints(FALSE)
+
+	var/turf/my_turf = get_turf(src)
+	for(var/thing in tracking_devices)
+		var/obj/item/gps/gps = locate(thing)
+		if(!istype(gps) || QDELETED(gps))
+			LAZYREMOVE(tracking_devices, thing)
+			LAZYREMOVE(showing_tracked_names, thing)
+			continue
+	
+		var/turf/gps_turf = get_turf(gps)
+		var/gps_tag = LAZYACCESS(showing_tracked_names, thing) ? gps.gps_tag : null
+		if(istype(gps_turf))
+			compass.set_waypoint("\ref[gps]", gps_tag, gps_turf.x, gps_turf.y, gps_turf.z, LAZYACCESS(tracking_devices, "\ref[gps]"))
+			if(can_track(gps) && my_turf && gps_turf != my_turf)
+				compass.show_waypoint("\ref[gps]")
+	
+	compass.rebuild_overlay_lists(update_compass_icon)
+
+/obj/item/gps/AltClick(mob/user)
+	toggletracking(user)
+
+/obj/item/gps/proc/toggletracking(mob/living/user)
+	if(!istype(user))
+		return
+
+	if(emped)
+		to_chat(user, SPAN_WARNING("\The [src] is busted!"))
+		return
+	
+	toggle_tracking()
+	if(tracking)
+		to_chat(user, SPAN_NOTICE("\The [src] is now tracking, and visible to other GPS devices."))
+	else
+		to_chat(user, SPAN_NOTICE("\The [src] is no longer tracking, or visible to other GPS devices."))
+
+/obj/item/gps/proc/toggle_tracking()
```

Comment:
Please don't have two procs named the same thing except one using normal style and one not.

Reply:
Merged.